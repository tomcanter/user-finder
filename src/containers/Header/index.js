// @flow

import React, {
  Component,
  PropTypes,
} from 'react';
import {
  StyleSheet,
} from 'aphrodite/no-important';
import Header from 'components/Header';
import Container from 'components/Container';
import connect from './connect';

type Props = {
  onSubmit: Function,
  searchQuery: string,
  searchTerm: string,
  searchForUser: Function,
};

export class HeaderContainer extends Component {

  static propTypes = {
    onSubmit: PropTypes.func.isRequired,
    searchQuery: PropTypes.string.isRequired,
    searchTerm: PropTypes.string.isRequired,
    searchForUser: PropTypes.func.isRequired,
  };

  static defaultProps = {
    searchTerm: '',
  };

  constructor(props: Props) {
    super(props);
    this.handleSearchUser(props.searchQuery);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.searchQuery !== this.props.searchQuery) {
      this.handleSearchUser(nextProps.searchQuery);
    }
  }

  handleSearchUser(searchQuery: string): void {
    if (!searchQuery) {return;}
    this.props.searchForUser(searchQuery);
  }

  render() {
    const {
      onSubmit,
      searchTerm,
    } = this.props;

    return (
      <Container rootStyle={styles.HeaderContainer}>
        <Header
          onSubmit={onSubmit}
          searchTerm={searchTerm}
        />
      </Container>
    );
  }

}

const styles = StyleSheet.create({
  HeaderContainer: {
    backgroundColor: '#f4f4f4',
    borderBottom: '1px solid #aaa',
    boxShadow: '0px 0px 2px 1px rgba(0, 0, 0, .2)',
  },
});

export default connect(HeaderContainer);
