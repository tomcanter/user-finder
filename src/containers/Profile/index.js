import React, {
  Component,
} from 'react';
import get from 'lodash/fp/get';
import {
  css,
  StyleSheet,
} from 'aphrodite/no-important';
import Container from 'components/Container';
import ProfileHeader from 'components/ProfileHeader';
import ItemList from 'components/ItemList';
import User from 'components/User';
import Repo from 'components/Repo';
import Loading from 'components/Loading';
import ErrorBoundary from 'components/ErrorBoundary';
import {viewport} from 'theme';
import 'suitcss-utils-flex/lib/flex-sm.css';
import 'suitcss-utils-size/lib/size-sm.css';
import 'suitcss-components-grid';
import connect from './connect';

type Props = {
  followerIds: Array<number>,
  followerIsPending: boolean,
  getProfile: Function,
  userIsPending: boolean,
  userProfile: Object,
  repoEntities: Object,
  repoIds: Array<number>,
  repoIsPending: boolean,
  userEntities: Object,
};

const getUsername = get('match.params.username');

export class ProfileContainer extends Component {

  props: Props;

  constructor(props: Props) {
    super(props);
    this.props.getProfile(getUsername(props));
  }

  componentWillReceiveProps(nextProps: Props) {
    const username = getUsername(this.props);
    const nextUsername = getUsername(nextProps);

    if (username !== nextUsername) {
      this.props.getProfile(nextUsername);
    }
  }

  renderProfile() {
    if (this.props.userIsPending) {
      return (
        <div className={css(styles.Profile_wrapLoading)}>
          <Loading />
        </div>
      );
    }

    const {
      repoIsPending,
      repoEntities,
      repoIds,
      followerIds,
      followerIsPending,
      userEntities,
    } = this.props;

    return (
      <div className={css(styles.Profile_container)}>
        <ErrorBoundary>
          <ProfileHeader {...this.props.userProfile} />
        </ErrorBoundary>
        <div className={`${css(styles.Profile_content)} Grid Grid--withGutter`}>
          <div className={`${css(styles.Profile_repos)} Grid-cell u-sm-size1of2`}>
            <h2 className={css(styles.Profile_contentTitle)}>Repositories</h2>
            <ErrorBoundary>
              <ItemList
                entities={repoEntities}
                ids={repoIds}
                isPending={repoIsPending}
                component={Repo}
              />
            </ErrorBoundary>
          </div>
          <div className="Grid-cell u-sm-size1of2">
            <h2 className={css(styles.Profile_contentTitle)}>Followers</h2>
            <ErrorBoundary>
              <ItemList
                entities={userEntities}
                ids={followerIds}
                isPending={followerIsPending}
                component={User}
              />
            </ErrorBoundary>
          </div>
        </div>
      </div>
    );
  }

  render() {
    return (
      <Container>
        {this.renderProfile()}
      </Container>
    );
  }

}

const styles = StyleSheet.create({
  Profile_wrapLoading: {
    marginTop: 25,
  },

  Profile_repos: {
    marginBottom: 20,

    [viewport.SM]: {
      marginBottom: 0,
    },
  },

  Profile_contentTitle: {
    fontSize: 22,
    fontWeight: 400,
    marginBottom: 25,
  },

  Profile_content: {
    paddingTop: 20,
  },

  Profile_container: {
    paddingTop: 15,
  },
});

export default connect(ProfileContainer);
