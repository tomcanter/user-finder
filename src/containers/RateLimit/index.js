import React from 'react';
import {
  css,
  StyleSheet,
} from 'aphrodite/no-important';
import 'suitcss-utils-flex/lib/flex.css';
import RateLimit from 'components/RateLimit';
import connect from './connect';

type Props = {
  coreLimit: ?number,
  coreRemaining: ?number,
  coreReset: ?string,
  searchLimit: ?number,
  searchRemaining: ?number,
  searchReset: ?string,
};

export function RateLimitContainer(props: Props) {
  const {
    coreLimit,
    coreRemaining,
    coreReset,
    searchLimit,
    searchRemaining,
    searchReset,
  } = props;

  return (
    <div className="u-flex u-flexAlignItemsCenter">
      <p className={css(styles.RateLimitContainer_label)}>API requests</p>
      <div
        className={`${css(styles.RateLimitContainer)} u-flex`}
      >
        <div className={css(styles.RateLimitContainer_item)}>
          <RateLimit
            label="Profile"
            remaining={coreRemaining}
            limit={coreLimit}
            reset={coreReset}
          />
        </div>
        <div className={css(styles.RateLimitContainer_item)}>
          <RateLimit
            label="Search"
            remaining={searchRemaining}
            limit={searchLimit}
            reset={searchReset}
          />
        </div>
      </div>
    </div>
  );
}

const styles = StyleSheet.create({
  RateLimitContainer: {
    backgroundColor: '#333',
    color: '#eee',
    padding: '5px 8px',
    borderRadius: 5,
  },

  RateLimitContainer_label: {
    fontSize: 14,
    marginRight: 10,
  },

  RateLimitContainer_item: {
    ':first-child': {
      marginRight: 5,
    },

    ':last-child': {
      marginLeft: 5,
    },
  },
});

export default connect(RateLimitContainer);
