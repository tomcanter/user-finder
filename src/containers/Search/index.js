// @flow

import React, {
  Component,
  PropTypes,
} from 'react';
import {
  StyleSheet,
} from 'aphrodite/no-important';
import SearchForm from 'components/SearchForm';
import SearchResults from 'components/SearchResults';
import Pagination from 'components/Pagination';
import SearchHeader from 'components/SearchHeader';
import Container from 'components/Container';
import connect from './connect';

type Props = {
  onSubmit: Function,
  pagination: Object,
  search: Object,
  search: string,
  searchUser: Function,
  totalResults: number,
  userEntities: Object,
  userIds: Array<number>,
};

export class SearchContainer extends Component {

  static propTypes = {
    onSubmit: PropTypes.func.isRequired,
    pagination: PropTypes.object,
    search: PropTypes.string.isRequired,
    searchTerm: PropTypes.string.isRequired,
    searchUser: PropTypes.func.isRequired,
    totalResults: PropTypes.number.isRequired,
    userEntities: PropTypes.object.isRequired,
    userIds: PropTypes.array.isRequired,
  };

  static defaultProps = {
    searchTerm: '',
    pagination: {},
  };

  constructor(props: Props) {
    super(props);
    this.handleSearchUser(props.search);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.search !== this.props.search) {
      this.handleSearchUser(nextProps.search);
    }
  }

  handleSearchUser(search: string): void {
    if (!search) {return;}
    this.props.searchUser(search);
  }

  renderSearchResults() {
    const {
      searchTerm,
      userIds,
      userEntities,
      totalResults,
    } = this.props;
    if (!searchTerm) {return null;}

    return (
      <SearchResults
        searchTerm={searchTerm}
        ids={userIds}
        entities={userEntities}
        total={totalResults}
      />
    );
  }

  renderPagination() {
    const {
      pagination,
      searchTerm,
    } = this.props;
    if (!searchTerm) {return null;}

    return (
      <Container rootStyle={styles.SearchContainer_pagination}>
        <Pagination {...pagination} />
      </Container>
    );
  }

  render() {
    const {
      onSubmit,
      searchTerm,
    } = this.props;

    return (
      <div>
        <Container rootStyle={styles.SearchContainer_header}>
          <SearchHeader>
            <SearchForm
              onSubmit={onSubmit}
              initialInputValue={searchTerm}
            />
          </SearchHeader>
        </Container>
        {this.renderPagination()}
        <Container noGutter={true}>
          {this.renderSearchResults()}
        </Container>
      </div>
    );
  }

}

const styles = StyleSheet.create({
  SearchContainer_pagination: {
    marginTop: 15,
    marginBottom: 15,
  },

  SearchContainer_header: {
    backgroundColor: '#f4f4f4',
    borderBottom: '1px solid #ddd',
  },
});

export default connect(SearchContainer);
