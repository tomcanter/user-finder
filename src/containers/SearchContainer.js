// @flow

import React, {
  Component,
  PropTypes,
} from 'react';
import SearchForm from '../components/SearchForm';
import SearchResults from '../components/SearchResults';
import connect from './SearchConnect';

type Props = {
  onSubmit: Function,
  searchUser: Function,
  userEntities: Object,
  userIds: Array<number>,
  totalResults: number,
  query: string,
};

export class SearchContainer extends Component {

  static propTypes = {
    onSubmit: PropTypes.func.isRequired,
    query: PropTypes.string.isRequired,
    searchUser: PropTypes.func.isRequired,
    totalResults: PropTypes.number.isRequired,
    userEntities: PropTypes.object.isRequired,
    userIds: PropTypes.array.isRequired,
  };

  constructor(props: Props) {
    super(props);
    this.handleSearchUser(props.query);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.query !== this.props.query) {
      this.handleSearchUser(nextProps.query);
    }
  }

  handleSearchUser(query: string): void {
    if (!query) {return;}
    this.props.searchUser({query});
  }

  render() {
    const {
      onSubmit,
      query,
      totalResults,
      userEntities,
      userIds,
    } = this.props;

    return (
      <div>
        <SearchForm
          onSubmit={onSubmit}
          initialInputValue={query}
        />
        <SearchResults
          query={query}
          ids={userIds}
          entities={userEntities}
          total={totalResults}
        />
      </div>
    );
  }

}

export default connect(SearchContainer);
