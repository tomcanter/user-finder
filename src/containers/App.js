// @flow

import React from 'react';
import {
  BrowserRouter as Router,
  Route,
} from 'react-router-dom';

import 'suitcss-base';
import SearchScreen from '../screens/SearchScreen';
import Profile from '../screens/Profile';

export default function App() {
  return (
    <Router>
      <div className="App">
        <Route exact={true} path="/" component={SearchScreen} />
        <Route exact={true} path="/:username" component={Profile} />
      </div>
    </Router>
  );
}
