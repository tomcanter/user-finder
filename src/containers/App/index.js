// @flow

import React from 'react';
import {
  HashRouter as Router,
  Redirect,
  Switch,
} from 'react-router-dom';
import SearchScreen from 'screens/Search';
import Profile from 'screens/Profile';
import DefaultLayout from 'screens/Default';

import 'suitcss-base';
import 'css/global.css';

export default function App() {
  return (
    <Router>
      <Switch>
        <DefaultLayout exact={true} path="/search" component={SearchScreen} />
        <DefaultLayout exact={true} path="/:username" component={Profile} />
        <Redirect to="/search" />
      </Switch>
    </Router>
  );
}
