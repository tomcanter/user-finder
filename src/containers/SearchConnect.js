// @flow

import {connect} from 'react-redux';
import get from 'lodash/fp/get';
import {searchUser} from '../store/Search/actions';

export function mapStateToProps(state: Object): Object {
  const userIds = get('search.userIds', state);
  return {
    userEntities: get('entities.users', state),
    userIds,
    totalResults: userIds.length,
  };
}

const mapDispatchToProps = {
  searchUser,
};

export default connect(mapStateToProps, mapDispatchToProps);
