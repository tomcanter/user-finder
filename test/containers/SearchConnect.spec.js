import {mapStateToProps} from '../../src/containers/SearchConnect';

describe('SearchConnect: mapStateToProps', () => {

  describe('when given the state', () => {
    it('should return relevant properties', () => {
      const state = {
        search: {
          result: [1, 2, 3],
        },
        entities: {
          users: {
            1: {id: 1},
            2: {id: 2},
          },
        },
      };
      expect(mapStateToProps(state)).toMatchSnapshot();
    });
  });

});
