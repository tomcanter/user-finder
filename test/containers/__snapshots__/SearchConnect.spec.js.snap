exports[`SearchConnect: mapStateToProps when given the state should return relevant properties 1`] = `
Object {
  "totalResults": 3,
  "userEntities": Object {
    "1": Object {
      "id": 1,
    },
    "2": Object {
      "id": 2,
    },
  },
  "userIds": Array [
    1,
    2,
    3,
  ],
}
`;
